### PR TITLE
bpf: convert majority of `bpf_elf_map` definitions to BTF map definitions 

### PR DESCRIPTION
**Pull request by ti-mo**
Implements part of: https://github.com/cilium/cilium/issues/15149.

Best reviewed per commit.

Note that these were converted by hand using some macros, so would like an extra set of eyes to verify before/after.

This notably omits `PROG_ARRAY` and `HASH_OF_MAPS` maps because there is some extra work that needs to be done for those. They will be addressed in a follow-up.

```release-note
bpf: convert majority of `bpf_elf_map` definitions to BTF map definitions
```

### CODE REVIEW

File: bpf/bpf_sock.c [hunk 1]
```diff
@@ -179,13 +179,13 @@ bool sock_proto_enabled(__u32 proto)
...
-	.size_key	= sizeof(struct ipv4_revnat_tuple),
-	.size_value	= sizeof(struct ipv4_revnat_entry),
-	.pinning	= PIN_GLOBAL_NS,
-	.max_elem	= LB4_REVERSE_NAT_SK_MAP_SIZE,
-};
+struct {
+	__uint(type, BPF_MAP_TYPE_LRU_HASH);
+	__type(key, struct ipv4_revnat_tuple);
+	__type(value, struct ipv4_revnat_entry);
+	__uint(pinning, 1);
```
Comment (pchaigno) [inline bpf/bpf_sock.c]: Shouldn't we use macro `LIBBPF_PIN_BY_NAME` here?

File: bpf/lib/lb.h [hunk 2]
```diff
@@ -175,14 +175,14 @@ struct bpf_elf_map __section_maps LB4_MAGLEV_MAP_OUTER = {
 #endif /* ENABLE_IPV4 */
 
 #ifdef ENABLE_SESSION_AFFINITY
-struct bpf_elf_map __section_maps LB_AFFINITY_MATCH_MAP = {
-	.type		= BPF_MAP_TYPE_HASH,
-	.size_key	= sizeof(struct lb_affinity_match),
-	.size_value	= sizeof(__u8), /* dummy value, map is used as a set */
```
Comment (pchaigno) [inline bpf/lib/lb.h:181]: Not related to this PR but I wonder if it would make sense to have a `typedef bool` for those `__u8` values, to clarify the semantics (instead of commenting on each).

File: bpf/lib/maps.h [hunk 3]
```diff
@@ -10,26 +10,31 @@
...
-	.max_elem	= METRICS_MAP_SIZE,
-	.flags		= CONDITIONAL_PREALLOC,
-};
 
 #ifndef SKIP_POLICY_MAP
-/* Global map to jump into policy enforcement of receiving endpoint */
+/*
+ * Global map to jump into policy enforcement of receiving endpoint
+ * TODO(timo): Convert to BTF map definition w/ static values declaration
+ * when iproute2 is no longer used.
```
Comment (pchaigno) [inline bpf/lib/maps.h]: Don't we already have the issue to track this?

File: bpf/lib/nodeport.h [hunk 4]
```diff
@@ -59,23 +59,23 @@ bpf_skip_nodeport(struct __ctx_buff *ctx)
...
-struct bpf_elf_map __section_maps NODEPORT_NEIGH6 = {
-	.type		= BPF_MAP_TYPE_LRU_HASH,
-	.size_key	= sizeof(union v6addr),		/* ipv6 addr */
-	.size_value	= sizeof(union macaddr),	/* hw addr */
-	.pinning	= PIN_GLOBAL_NS,
-	.max_elem	= NODEPORT_NEIGH6_SIZE,
-};
+struct {
+	__uint(type, BPF_MAP_TYPE_LRU_HASH);
+	__type(key, union v6addr);		/* ipv6 addr */
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: One tab too many before the comment here.

File: bpf/tests/prog_test/prog_test.go [hunk 5]
```diff
@@ -436,20 +337,29 @@ func testCt4Rst(spec *ebpf.Collection) error {
...
 }
 
+func modifyMapSpecs(spec *ebpf.CollectionSpec) {
+	for _, m := range spec.Maps {
+		// Clear pinning flag on all Maps, keep this test self-contained.
+		m.Pinning = 0
+
+		// Drain Extra section of legacy bpf_elf_map definitions. The library
+		// rejects any bytes left over in Extra on load.
+		io.Copy(io.Discard, &m.Extra)
```
Comment (pchaigno) [inline bpf/tests/prog_test/prog_test.go:347]: Probably a naive Go question, but why is this needed? Isn't it a noop statement?

File: bpf/include/bpf/loader.h [hunk 6]
```diff
@@ -8,6 +8,7 @@
 
 #define __uint(name, val) int(*(name))[val]
 #define __type(name, val) typeof(val) *(name)
+#define __array(name, val) typeof(val) *(name)[]
```
Comment (pchaigno) [inline bpf/include/bpf/loader.h:11]: Commit title can be reduce to fit the max length by changing to:
```diff
-  test/bpf: elf-demo - convert CALLS_MAP in test suite to BTF map definition
+ test/bpf: elf-demo - convert CALLS_MAP to BTF map definition
```
Given the tag at the beginning of the commit title, I don't think we'll lose meaning.

**Review by pchaigno — APPROVED** [review]
LGTM, only minor comments below.

I'm not super confident in my review of the Maglev inner map changes, but I see @borkmann is marked as a reviewer anyway.

File: bpf/tests/prog_test/prog_test.go (hunk 5 above)
Comment (ti-mo) [reply bpf/tests/prog_test/prog_test.go:347]: Yup, this drains `m.Extra`'s buffer as described in the comment, advancing the buffer's read pointer. Not a no-op. :wink: 

If the lib finds any non-zero bytes in `m.Extra`, (so, non-zero fields past the `flags` field in `bpf_elf_map`) map creation is explicitly rejected (by the lib, not the kernel). We chose to do it this way since it means parts of the map definition remains unparsed, so this shouldn't fail silently.

File: bpf/lib/maps.h (hunk 3 above)
Comment (ti-mo) [reply bpf/lib/maps.h]: Sure, will remove. :+1: We should indeed prioritize issues over TODOs, not much point in having it there. :)

File: bpf/tests/prog_test/prog_test.go (hunk 5 above)
Comment (pchaigno) [reply bpf/tests/prog_test/prog_test.go:347]: > advancing the buffer's read pointer

But we don't reuse that `m.Extra` anywhere, do we?

Comment (ti-mo) [reply bpf/tests/prog_test/prog_test.go:347]: It's [checked by the lib at map creation time](https://github.com/cilium/ebpf/blob/af3a2fe217c38e9a4e10501ff578fefe1317bd3d/map.go#L311-L317), so it needs to be drained.

Comment (pchaigno) [reply bpf/tests/prog_test/prog_test.go:347]: Aah :+1: Thanks for the explanation! :pray:

File: bpf/bpf_sock.c (hunk 1 above)
Comment (ti-mo) [reply bpf/bpf_sock.c]: We probably should, but it's the only enum value that's accepted in this field for now.

I'm a bit skeptical about the semantics. Pinning 'by name' as opposed to what? Also, iproute2 used to take `PIN_GLOBAL_NS` and `PIN_LOCAL_NS` here, but with libbpf+iproute2 only a value of 1 is accepted, which iproute2 assumes to mean `PIN_GLOBAL_NS`, since it pins in `/sys/fs/bpf/tc/globals/*`.

Don't feel strongly either way, I should probably change it to stick to the convention.

Edit: done.

File: bpf/lib/lb.h (hunk 2 above)
Comment (ti-mo) [reply bpf/lib/lb.h:181]: Good idea for a follow-up? I'm also not sure if it's really necessary.

File: bpf/lib/ids.h [hunk 7]
```diff
@@ -6,8 +6,6 @@
 
 #define CILIUM_MAP_POLICY	1
 #define CILIUM_MAP_CALLS	2
-#define CILIUM_MAP_MAGLEV4	3
-#define CILIUM_MAP_MAGLEV6	4
```
Comment (qmonnet) [inline bpf/lib/ids.h:10]: Nit: I think it should be safe to change `CILIUM_MAP_CUSTOM_CALLS` below to 3 instead, now that macros using values 3 and 4 are gone?

File: pkg/maps/lbmap/maglev_inner_map.go [hunk 8]
```diff
@@ -27,9 +27,9 @@ type MaglevInnerVal struct {
 }
 
 // newMaglevInnerMapSpec returns the spec for a maglev inner map.
-func newMaglevInnerMapSpec(name string, tableSize uint32) *ebpf.MapSpec {
+func newMaglevInnerMapSpec(tableSize uint32) *ebpf.MapSpec {
 	return &ebpf.MapSpec{
-		Name:       name,
+		Name:       "maglev_inner",
```
Comment (qmonnet) [inline pkg/maps/lbmap/maglev_inner_map.go]: Should we keep the `cilium_` prefix for more clarity? In case someone logs maps creation for example.

**Review by qmonnet — APPROVED** [review]
Nice work! Two nits below (non-blocking).

As a side note, it's not often that we touch so many `bpf/` files, and you could take this chance to update the copyright year in most of the files you changed.

File: bpf/lib/ids.h (hunk 7 above)
Comment (ti-mo) [reply bpf/lib/ids.h:10]: True, it's not a kernel property, so as long as the section names correspond to the map's `.id` within the same ELF, we should be good.

**Review by borkmann — APPROVED** [review]
